File: kat/src/test/java/plus/kat/SpareTest.java
package plus.kat;

import lombok.*;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * @author kraity
 */
public class SpareTest {

    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    static class User {
        private int id;
        private String name;
    }

    @Test
    public void test_context() {
        Spare<User> spare =
            Spare.of(User.class);

        assertNotNull(spare);
        assertNotNull(spare.getContext());
    }

    @Test
    public void test_apply() {
        Spare<User> spare =
            Spare.of(User.class);

        assertNotNull(spare.apply());
    }

    @Test
    public void test_apply_type() {
        Spare<User> spare =
            Spare.of(User.class);

        User user0 = spare.apply();
        assertNotNull(user0);
        assertEquals(0, user0.id);

        User user1 = spare.apply(1, "kraity");
        assertNotNull(user1);
        assertEquals(1, user1.id);
        assertEquals("kraity", user1.name);

        Object[] nil = null;
        assertThrows(NullPointerException.class, () -> spare.apply(nil));
    }
}

File: kat/src/test/kotlin/plus/kat/SpareTest.kt
package plus.kat

import plus.kat.*
import plus.kat.actor.*
import plus.kat.entity.*

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*

class SpareKtTest {

    @Magus
    class User(
        @Magic("id")
        var id: Int = 1,

        @Magic("name")
        var name: String = "kraity"
    )

    @Test
    fun test_marker() {
        val spare = spare<User>()

        val user0 = spare.read(
            Flow.of("{}")
        )
        assertNotNull(user0)
        assertEquals(1, user0.id)
        assertEquals("kraity", user0.name)
        assertEquals("""{name="kraity",id=1}""", spare.write(user0).use { it.toString() })

        val user2 = spare<User>().read(
            Flow.of("{id=2}")
        )
        assertNotNull(user2)
        assertEquals(2, user2.id)
        assertEquals("kraity", user2.name)
        assertEquals("""{name="kraity",id=2}""", spare.write(user2).use { it.toString() })
    }
}
